### DISCUSSION TITLE
Disable Alt+` on Windows

### DISCUSSION THREAD

**Discussion by niwazukihon** [google/mozc#1147] [Q&A] [answered]
Hi, I'd like to disable the `` Alt+` `` hotkey to switch between hiragana and direct input on Windows, because I want to use that hotkey for other things. It's not in the keymap settings. I found #395 and saw the quoted code. I deleted the code

```cc
 {kTipPreservedKey_Kanji,
     {VK_OEM_3, TF_MOD_ALT},
     VK_OEM_3,
     &kTipKeyTilde[0],
     std::size(kTipKeyTilde) - 1},
```

from [src/win32/tip/tip_text_service.cc](https://github.com/google/mozc/blob/master/src/win32/tip/tip_text_service.cc#L328-L333), but it doesn't work.

**Comment by yukawa** [comment]
OK, on my Windows 11 24H2 environment, the behavior observed from the IME (e.g. Mozc) has changed from what I described in #395.

As of Windows 11 24H2, I think you can achieve your goal as follows.

1. Open Mozc property. (e.g. right click the Mozc mode icon on the task bar then select "Property")
2. Click "Customize" in the "keymap style section"
![image](https://github.com/user-attachments/assets/60a89a81-7d90-4b3f-bdbb-1afce5da5b26)
3. Remove all the 4 entries associated with "Hankaku/Zenkaku"
![image](https://github.com/user-attachments/assets/bf8c4678-78e3-4e7d-9116-c7b4677d5d47)

**Reply by yukawa** [reply, marked as answer]
Note that here are what I can observe in my local environment.

1. Open a notepad app.
2. Down <kbd>Alt</kbd> key.
3. Down <kbd>\`</kbd> key. Mozc receives `VK_DBE_DBCSCHAR` down.
4. Up <kbd>\`</kbd> key. Mozc receives `VK_KANJI` up.
5. Down <kbd>\`</kbd> key. Mozc receives `VK_DBE_DBCSCHAR` down.
6. Up <kbd>\`</kbd> key. Mozc receives `VK_KANJI` up.
7. Up  <kbd>Alt</kbd> key.

As you can see, what Mozc is receiving are `VK_DBE_DBCSCHAR` and `VK_KANJI`. Such a conversion is done at the OS layer.

In this scenario, Mozc basically ignores key up events so let's ignore `VK_KANJI` at the step 4 and step 6.
When Mozc receives `VK_DBE_DBCSCHAR`, it will be internally interpreted as "Hankaku/Zenkaku" as follows.
https://github.com/google/mozc/blob/4c7f0482f7615a1c343593f465e9b1363fbf6163/src/win32/base/keyevent_handler.cc#L316

This is why you need to modify the key bindings associated with "Hankaku/Zenkaku".

**Reply by niwazukihon** [reply]
Thank you for the reply. I removed the 4 entries, but the hotkey I mapped in autohotkey still doesn't work when mozc is activated. But from your comment, I guess you meant that the OS gives a special treatment of `` alt+` `` when using Japanese keyboard, and when I removed all 4 entries, essentially mozc doesn't do anything with what comes out of the OS.

Then, by looking into the ahk key history, I found that it seems ahk is receiving `vk19` (`VK_KANJI`) (and `vkE8`). Holding it down keeps sending `vk19`. I don't see `vkF4`. So, I tried this script

```ahk
<!`:: Home
<!vk19:: Home
```

And it works. In fact, it even works without removing those 4 entries.

**Reply by yukawa** [reply]
> Then, by looking into the ahk key history, I found that it seems ahk is receiving `vk19` (`VK_KANJI`) (and `vkE8`). Holding it down keeps sending `vk19`. I don't see `vkF4`.

Ah, you are right. I've also confirmed that `VK_KANJI` is received not only for key up but also for key down.

The reason why I thought `VK_DBE_DBCSCHAR` was sent was because of the following special mapping rule.
https://github.com/google/mozc/blob/b7faa84f2fea1332643ee7169e914854096bad67/src/win32/tip/tip_text_service.cc#L334-L342

So the system was indeed sending `VK_KANJI`, but the above rule defined in `tip_text_service.cc` further remapped it to `VK_DBE_DBCSCHAR`.
